perf(core): avoid storing LView in __ngContext__

crisbeto opened the pull request:
These changes combine #41358 and #41894.

Currently we save a reference to an `LView` on most DOM nodes created by Angular either by saving the `LView` directly in the `__ngContext__` or by saving the `LContext` which has a reference to the `LView`. This can be a problem if the DOM node is retained in memory, because the `LView` has references to all of the child nodes of the view, as well as other internal data structures.

Previously we tried to resolve the issue by clearing the `__ngContext__` when a node is removed (see https://github.com/angular/angular/pull/36011), but we decided not to proceeed, because it can slow down destruction due to a megamorphic write.

These changes aim to address the issue while reducing the performance impact by assigning a unique ID when an `LView` is created and adding it to `__ngContext__`. All active views are tracked in a map where their unique ID is used as the key. We don't need to worry about leaks within that map,  because `LView`s are an internal data structure and we have complete control over when they are  created and destroyed.

Fixes #41047.

crisbeto commented:
@mgechev last time we tried to land these changes there were also some concerns about the devtools. Is this still something we need to account for?

AndrewKushnir commented:
> last time we tried to land these changes there were also some concerns about the devtools. Is this still something we need to account for?

// cc @AleksanderBodurri

AleksanderBodurri commented:
@AndrewKushnir thanks for the heads up

@crisbeto

In Angular DevTools the value of `__ngContext__` is used in a few polyfill APIs that are only used for older versions Angular. Since this is targeting a major release I'm confident that these polyfills would not run for any Angular version that has this change.

Other than those polyfills, the existence of `__ngContext__` is used in a few places to determine if an Angular app has Ivy enabled. This should be fine since these usages are only checking for the existence of `__ngContext__` rather than its value. 

I built DevTools locally from this branch and everything is working as expected. Awesome PR 😄

mgechev commented:
Thanks for the heads up! As @AleksanderBodurri said we are good. Our check is `__ngContext__ !== 'undefined'` so we should be safe.

crisbeto commented:
[Passing TGP.](test/OCL:427953245:BASE:428969986:1644998316035:8a534487)

crisbeto committed:
perf(core): avoid storing LView in __ngContext__

These changes combine #41358 and #41894.

Currently we save a reference to an `LView` on most DOM nodes created by Angular either by saving the `LView` directly in the `__ngContext__` or by saving the `LContext` which has a reference to the `LView`. This can be a problem if the DOM node is retained in memory, because the `LView` has references to all of the child nodes of the view, as well as other internal data structures.

Previously we tried to resolve the issue by clearing the `__ngContext__` when a node is removed (see https://github.com/angular/angular/pull/36011), but we decided not to proceeed, because it can slow down destruction due to a megamorphic write.

These changes aim to address the issue while reducing the performance impact by assigning a unique ID when an `LView` is created and adding it to `__ngContext__`. All active views are tracked in a map where their unique ID is used as the key. We don't need to worry about leaks within that map,  because `LView`s are an internal data structure and we have complete control over when they are  created and destroyed.

Fixes #41047.

AndrewKushnir reviewed (approved):
👍

alxhub commented:
This PR was merged into the repository by commit 071c8af8ba89163270f17657c0c7f1c944f186f7.
